osu.Game.Tournament/Screens/StablePathSelectScreen.cs

```diff
@@ -0,0 +1,173 @@
+                                {
+                                    new FillFlowContainer
+                                    {
+                                        Anchor = Anchor.Centre,
+                                        Origin = Anchor.Centre,
+                                        Direction = FillDirection.Horizontal,
+                                        Spacing = new Vector2(20),
+                                        Children = new Drawable[]
+                                        {
+                                            new TriangleButton
+                                            {
+                                                Anchor = Anchor.Centre,
+                                                Origin = Anchor.Centre,
+                                                Width = 300,
+                                                Text = "Select stable path",
+                                                Action = () => ChangePath(storage)
+                                            },
+                                            new TriangleButton
+                                            {
+                                                Anchor = Anchor.Centre,
+                                                Origin = Anchor.Centre,
+                                                Width = 300,
+                                                Text = "Auto detect",
+                                                Action = AutoDetect
+                                            },
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    },
+                },
+                new BackButton
+                {
+                    Anchor = Anchor.BottomLeft,
+                    Origin = Anchor.BottomLeft,
+                    State = { Value = Visibility.Visible },
+                    Action = () => sceneManager?.SetScreen(typeof(SetupScreen))
+                }
+            });
+        }
+
+        protected virtual void ChangePath(Storage storage)
+        {
+            var target = directorySelector.CurrentDirectory.Value.FullName;
+            var fileBasedIpc = ipc as FileBasedIPC;
+            Logger.Log($"Changing Stable CE location to {target}");
+
+            if (!FileBasedIPC.CheckExists(target))
+            {
+                overlay = new DialogOverlay();
+                overlay.Push(new IPCErrorDialog("This is an invalid IPC Directory", "Select a directory that contains an osu! stable cutting edge installation and make sure it has an empty ipc.txt file in it."));
+                AddInternal(overlay);
+                Logger.Log("Folder is not an osu! stable CE directory");
+                return;
+            }
+
+            fileBasedIpc?.SaveStableConfig(target);
+            fileBasedIpc?.LocateStableStorage();
```

Comment:
This is calling three things in sequence on `fileBasedIpc`, which means that this snippet of code is a good candidate for encapsulating in a method. Better yet, if you exposed a method in `FileBasedIPC` like:

```csharp
public bool SetIPCLocation(string path);
```

then you could probably (a) get rid of the strange `string` return type and config lookup case in `getStablePath()`, *and* (b) you could probably reuse *that same* method inside of the stable path lookup code.